### PR TITLE
remove-{application,unit}: introduce --destroy-storage flag

### PR DESCRIPTION
**Pull request by axw**
Add a --destroy-storage flag to the remove-unit and
remove-application commands. When this flag is
specified, the attached storage will be destroyed,
rather than detached.

---

We introduce state.ModelOperation, implementations of
which encapsulate the logic for building operations
for making changes to a model. Two implementations
are provided, for destroying units and applications.
Separating the building of operations from their
execution allows for customisation of the operations.
We could alternatively use parameter structs, as we
do elsewhere, but the approach taken should enable
us to work towards composable operations, as described
below.

For now, operations can only be executed in isolation,
via state.State.ApplyOperation. This is due to several
limitations:

 - It is unsafe to compose mgo/txn ops arbitrarily.
   We currently compose mgo/txn ops across collections,
   which is unsafe, but the changes are made in a way
   as to limit the damage of interrupted transactions,
   and these static compositions are tested.
 - There is an upper limit on how many operations can
   be safely included in a mgo/txn transaction, which
   we do sometimes reach.

Eventually, if/when we move to an alternative replication
mechanism such as Raft, then we can look at making these
operations composable. For example, it would make sense
to compose hypothetical AddApplication and Addunit
operations, making the "NumUnit" logic in AddApplication
redundant. This would enable us to move more of the
business logic out of the state package, keeping it
focused on model-checking.

### CODE REVIEW

File: api/application/client.go [hunk 1]
```diff
@@ -319,33 +319,44 @@ func (c *Client) DestroyUnitsDeprecated(unitNames ...string) error {
 	return c.facade.FacadeCall("DestroyUnits", params, nil)
 }
 
+// DestroyUnitsParams contains parameters for the DestroyUnits API method.
+type DestroyUnitsParams struct {
```
Comment (rogpeppe) [inline api/application/client.go]: This doesn't seem related to the PR description. Maybe better to keep the PR more focused?

File: state/modeltransaction.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	ops []ModelTransactionOperation
+}
+
+// ModelTransactionOperation is a high-level transaction operation,
+// encapsulating the logic required to apply a change to a model.
+type ModelTransactionOperation interface {
+	// Build builds the low-level database transaction operations required
+	// to apply the change. If the transaction operations fail (e.g. due
+	// to concurrent changes), then Build may be called again. The attempt
+	// number can be obtained from the ModelTransactionContext.
```
Comment (rogpeppe) [inline state/modeltransaction.go]: Perhaps document the special errors that Build can return (for example, I see a test for ErrNoOperations below) ?

File: api/application/client.go [hunk 3]
```diff
@@ -369,32 +380,42 @@ func (c *Client) DestroyDeprecated(application string) error {
 	return c.facade.FacadeCall("Destroy", params, nil)
 }
 
+type DestroyApplicationsParams struct {
```
Comment (wallyworld) [inline api/application/client.go]: struct comment?

Comment (axw) [reply api/application/client.go]: done

File: apiserver/params/params.go [hunk 4]
```diff
@@ -442,16 +442,56 @@ type AddApplicationUnits struct {
...
+	// NOTE(axw) the JSON tag is "entities" for compatibility
+	// with the Entities type, which this replaces for newer
+	// versions of DestroyUnit.
+	Units []DestroyUnitParams `json:"entities"`
+}
+
+// DestroyUnitParams holds parameters for the Application.DestroyUnit call.
+type DestroyUnitParams struct {
+	// UnitTag holds the tag of the unit to destroy.
+	UnitTag string `json:"tag"`
```
[the remark this answers is not in the document]
Comment (axw) [reply apiserver/params/params.go]: changed to unit-tag, and application-tag for the application equivalent

**Comment by axw** [comment]
$$merge$$

**Comment by axw** [comment]
$$merge$$

**Comment by axw** [comment]
@pmatulis FYI, this PR adds a --destroy-storage flag to the remote-unit and remote-application commands. By default those commands will detach the storage attached to the unit(s), but leave it alive in the model. By supplying the flag, the storage will be detached and then destroyed.

**Comment by pmatulis** [comment]
@axw Ack :octopus:
